src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1754,3 +1729,192 @@ Node TheoryStringsRewriter::collectConstantStringAt( std::vector< Node >& vec, u
+              removeComponent = true;
+            }else{
+              //check how much overlap there is
+              // This is used to partially strip off the endpoint
+              // e.g. str.contains( str.++( "abc", x ), str.++( "cd", y ) ) --> str.contains( str.++( "c", x ), str.++( "cd", y ) )
+              overlap = r==0 ? s.overlap(t) : t.overlap(s);
+            }
+          }else{
+            Assert( ret<s.size() );
+            // can strip off up to the find position
+            // e.g. str.contains( str.++( "abc", x ), str.++( "b", y ) ) --> str.contains( str.++( "bc", x ), str.++( "b", y ) )
+            overlap = s.size()-ret;
+          }
+        }else if( n2[index1].getKind()==kind::STRING_ITOS ){
+          const std::vector<unsigned>& svec = s.getVec();
+          // can remove up to the first occurrence of a non-digit
+          for( unsigned i=0; i<svec.size(); i++ ){
+            unsigned sindex = r==0 ? i : svec.size()-i;
+            if(String::isDigit(svec[sindex])) {
+              break;
+            }else{
+              // e.g. str.contains( str.++( "a", x ), str.to.int(y) ) --> str.contains( x, str.to.int(y) )
+              overlap--;
+            }
+          }
+        }else{
+          //inconclusive
+        }
+        // process the overlap
+        if( overlap<s.size() ){
+          changed = true;
+          if( overlap==0 ){
+            removeComponent = true;
+          }else{
+            // can drop the prefix (resp. suffix) from the first (resp. last) component
+            if( r==0 ){
+              nb.push_back( NodeManager::currentNM()->mkConst(s.prefix(overlap)) );
+              n1[index0] = NodeManager::currentNM()->mkConst(s.suffix(overlap));
+            }else{
+              ne.push_back( NodeManager::currentNM()->mkConst(s.suffix(overlap)) );
+              n1[index0] = NodeManager::currentNM()->mkConst(s.prefix(overlap));
+            }
+          }
+        }
+      }else if( n1[index0].getKind()==kind::STRING_ITOS ){
+        if( n2[index1].isConst() ){
+          CVC4::String t = n2[index1].getConst<String>();
+          const std::vector<unsigned>& tvec = t.getVec();
+          removeComponent = n1.size()>1;
+          // if n1.size()==1, then if n2[index1] contains a non-digit, we can drop the entire component
+          //    e.g. str.contains( str.to.int(x), "123a45") --> false
+          // if n1.size()>1, then if n2[index1] does not contain a digit, we can drop the entire component
+          //    e.g. str.contains( str.++( str.to.int(x), y ), "abc") --> str.contains( y, "abc" )
+          for( unsigned i=0; i<tvec.size(); i++ ){
+            if(String::isDigit(tvec[i])) {
+              if(n1.size()>1) {
+                removeComponent = false;
+                break;
+              }
+            }else{
+              if(n1.size()==1) {
+                removeComponent = true;
+                return true;
```

Comment:
Why return here? We should signal that containment is not possible. It should be ok to use `break` here, so `n1` would be empty and since we know that `n2` is not, we can be sure that it will be rewritten to `false`.

Reply:
This has been refactored now so that its more clear, I think you're right that break was the more sensible thing to do here